eZ/Publish/Core/MVC/Symfony/Routing/UrlWildcardRouter.php

```diff
@@ -0,0 +1,210 @@
+use Psr\Log\LoggerInterface;
+use Symfony\Cmf\Component\Routing\ChainedRouterInterface;
+use Symfony\Cmf\Component\Routing\RouteObjectInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Routing\Exception\ResourceNotFoundException;
+use Symfony\Component\Routing\Exception\RouteNotFoundException;
+use Symfony\Component\Routing\Matcher\RequestMatcherInterface;
+use Symfony\Component\Routing\RequestContext;
+use Symfony\Component\Routing\RouteCollection;
+use Symfony\Component\Routing\Route as SymfonyRoute;
+
+class UrlWildcardRouter implements ChainedRouterInterface, RequestMatcherInterface
+{
+    public const URL_ALIAS_ROUTE_NAME = 'ez_urlalias';
+
+    /**
+     * @var \eZ\Publish\API\Repository\URLWildcardService
+     */
+    private $wildcardService;
+
+    /**
+     * @var \eZ\Publish\Core\MVC\Symfony\Routing\Generator\UrlAliasGenerator
+     */
+    private $generator;
+
+    /**
+     * @var \Symfony\Component\Routing\RequestContext
+     */
+    private $requestContext;
+
+    /**
+     * @var \Psr\Log\LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * @param \eZ\Publish\API\Repository\URLWildcardService $wildcardService
+     * @param \eZ\Publish\Core\MVC\Symfony\Routing\Generator\UrlAliasGenerator $generator
+     * @param \Symfony\Component\Routing\RequestContext $requestContext
+     * @param \Psr\Log\LoggerInterface $logger
+     */
+    public function __construct(
+        URLWildcardService $wildcardService,
+        UrlAliasGenerator $generator,
+        RequestContext $requestContext,
+        LoggerInterface $logger
+    ) {
+        $this->wildcardService = $wildcardService;
+        $this->generator = $generator;
+        $this->requestContext = $requestContext;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @param Request $request
+     *
+     * @return array An array of parameters
+     *
+     * @throws \Symfony\Component\Routing\Exception\ResourceNotFoundException
+     */
+    public function matchRequest(Request $request): array
+    {
+        $requestedPath = $request->attributes->get('semanticPathinfo', $request->getPathInfo());
+
+        try {
+            /** @var UrlWildcard $urlWildcard */
+            $urlWildcard = $this->getUrlWildcard($requestedPath);
+            $path = $urlWildcard->destinationUrl;
+            $redirect = $urlWildcard->forward;
+        } catch (NotFoundException $e) {
+            try {
+                /** @var URLWildcardTranslationResult $urlWildcardTranslationResult */
+                $urlWildcardTranslationResult = $this->getUrlWildcardTranslationResult($requestedPath);
+                $path = $urlWildcardTranslationResult->uri;
+                $redirect = $urlWildcardTranslationResult->forward;
+            } catch (NotFoundException $e) {
+                throw new ResourceNotFoundException($e->getMessage(), $e->getCode(), $e);
+            }
+        }
+
+        if (isset($this->logger)) {
```

Comment:
This expression is always true. Ref: https://github.com/ezsystems/ezpublish-kernel/pull/2609/files#diff-f739df759a10b5a8df9a6c0930145fcbR65